src/twisted/protocols/_smb/ntlm.py

```diff
@@ -0,0 +1,380 @@
+AV_TREE_NAME = 0x0005
+AV_FLAGS = 0x0006
+AV_TIMESTAMP = 0x0007
+AV_SINGLE_HOST = 0x0008
+AV_TARGET_NAME = 0x0009
+AV_CHANNEL_BINDINGS = 0x000A
+
+SERVER_VERSION = (6, 1, 1)
+# major version 6.1 = Vista, roughly speaking what this emulates
+
+MAGIC = b'NTLMSSP\0'
+
+HeaderType = base.nstruct("magic:8s packet_type:I")
+
+NegType = base.nstruct("""flags:L domain_len:H domain_max_len:H
+    domain_offset:L
+    workstation_len:H workstation_max_len:H workstation_offse:L
+    v_major:B v_minor:B v_build:H reserved:3s v_protocol:B""")
+
+AuthType = base.nstruct("""
+        lmc_len:H lmc_maxlen:H lmc_offset:I
+        ntc_len:H ntc_maxlen:H ntc_offset:I
+        domain_len:H domain_maxlen:H domain_offset:I
+        user_len:H user_maxlen:H user_offset:I
+        workstation_len:H workstation_max_len:H workstation_offset:I
+        ersk_len:H ersk_maxlen:H ersk_offset:I
+        flags:I
+        v_major:B v_minor:B v_build:H reserved:3s v_protocol:B
+        mic:16s""")
+
+NtParts = base.nstruct("""response:16s resp_type:B hi_resp_type:B
+reserved:6s time:Q client_challenge:8s reserved2:4s""")
+
+ChallengeType = base.nstruct("""
+target_len:H target_max_len:H target_offset:I
+flags:I challenge:8s reserved:8s
+targetinfo_len:H targetinfo_max_len:H targetinfo_offset:I
+v_major:B v_minor:B v_build:H reserved2:3s v_protocol:B
+""")
+
+
+
+class NTLMManager(object):
+    """
+    manage the NTLM subprotocol
+
+    @ivar credential: the user cred, available after the AUTH token received
+                      None prior to this
+    @type credential: L{IUsernameHashedPassword}
+    """
+    def __init__(self, domain):
+        """
+        @param domain: the server NetBIOS domain
+        @type domain: L{str}
+        """
+        self.credential = None
+        self.flags = DEFAULT_FLAGS
+        self.server_domain = domain
+
+    def receiveToken(self, token):
+        """
+        receive client token once unpacked from overlying protocol
+
+        @type token: L{bytes}
+        """
+        self.token = token
+        if len(token) < 36:
+            log.debug("{tok}", tok=repr(token))
+            raise base.SMBError("token too small")
+        hdr = HeaderType(token)
+        if hdr.magic != MAGIC:
+            log.debug("{tok}", tok=repr(token[:16]))
+            raise base.SMBError("No valid NTLM token header")
+        try:
+            getattr(self, 'ntlm_' + NTLM_MESSAGES[hdr.packet_type])(hdr.buffer)
+        except IndexError:
+            raise base.SMBError("invalid message %d" % hdr.packet_type)
+
+    def ntlm_invalid(self, data):
+        raise base.SMBError("invalid message id 0")
+
+    def ntlm_challenge(self, data):
+        raise base.SMBError("invalid to send NTLM challenge to a server")
+
+    def ntlm_negotiate(self, data):
+        neg = NegType(data)
+        flags = flags2set(neg.flags)
+        log.debug("NTLM NEGOTIATE")
+        log.debug("--------------")
+        log.debug("Flags           %r" % list(flags))
```

Comment:
This is not how you use Twisted's logger. The string being passed is a *format* string, not a message.  So this should be:

`log.debug("Flags {flags}", flags=list(flags))`